### PR TITLE
🤖 Add authors and contributors to Practice Exercises

### PR DESCRIPTION
**Pull request by ErikSchierboom**
_If you got tagged in this PR, it is because you are being credited for work you've done in the past on Exercism here, and we want to ensure you don't miss out on the recognition you deserve 🙂_

---

With v3, we've introduced the idea of exercise authors and contributors. This information is stored in the exercises' `.meta/config.json` files (see [the docs](https://github.com/exercism/docs/blob/main/building/tracks/practice-exercises.md#file-metaconfigjson)).

Concept Exercises, which are all new, already contain authors and contributors. Practice Exercises though are missing this information. In this PR, we attempt to populate the authors and contributors of Practice Exercises based on their commit history.

## Maintainer TODO list

These are the things you, the maintainers, should do:

- [ ] Check the authors and contributors, adding/removing users where the data is missing or incorrect
- [ ] Double-check any renamed Practice Exercises
- [ ] Check for Practice Exercises with missing authors, which are most likely due to the author not being on GitHub anymore 
- [ ] Add additional authors if you know that a Practice Exercise has actually been created by more than one author 

**For clarity, authors should be the people who originally created the exercise on this track. Contributors should be people who have substantially contributed to that exercise. PRs for typos or multiple-exercise PRs do not automatically mean someone should be considered a contributor to that exercise. Those non-exercise-specific contributions will get recognition through Exercism's reputation system in v3.**

If you find something needs updating, just push a new commit to this PR.

## Automatic Merging

We will automatically merge this PR before we launch v3, but will give the maximum time we can to maintainers to review it first.

---

_The rest of this issue explains how this PR has been generated. You do not **need** to read it._

## Algorithm

We start out by looking at the current Practice Exercises. For each Practice Exercise, we'll look at the commit history of its files to see which commits touched that Practice Exercise. How renames are handled is discussed in the "Renames" section later in this document. 

Any commit that we can associate with a Practice Exercise is used to determine authorship/contributorship. Here is how we assign authorship/contributorship:

### Authors

1. Find the Git commit author of the oldest commit linked to that Practice Exercise.
2. Find the GitHub username for the Git commit author of that commit
3. Add the GitHub username of the Git commit author to the `authors` key in the `.meta/config.json` file

### Contributors

1. Find the Git commit authors and any co-authors of all but the oldest commit linked to that Practice Exercise. If there is only one commit, there won't be any contributors.
1. Exclude the Git commit author and any co-authors of the oldest commit from the list of Git commit authors (an author is _not_ also a contributor) 
2. Find the GitHub usernames for the Git commit authors and any co-authors of those commits
3. Add the GitHub usernames of the Git commit authors and any co-authors to the `contributor` key in the `.meta/config.json` file

We're using the GitHub GraphQL API to find the username of a commit author and any co-authors. In some cases though, a username cannot be found for a commit (e.g. due to the user account no longer existing), in which case we'll skip the commit. 

## Renames

There are a small number of Practice Exercises that might have been renamed at some point. You can ask Git to "follow" a file over its renames using `git log --follow <file>`, which will also return commits made before renames. Unfortunately, Git does not store renames, it just stores the contents of the renamed files and tries to guess if a file was renamed by looking at its contents. This _can_ (and will) lead to false positives, where Git will think a file has been renamed whereas it hasn't. As we don't want to have incorrect authors/contributors for exercises, we're ignoring renames. The only exception to this are known exercise renames:

- `bracket-push` was renamed to `matching-brackets`
- `retree` was renamed to `satellite`
- `resistor-colors` was renamed to `resistor-color-duo`
- `kindergarden-garden` was renamed to `kindergarten-garden`

If you know of a rename of a Practice Exercise, please check to see if its authors and contributors are correctly set.

## Exclusions

There are some commits that we skip over, which thus don't influence the authors/contributors list:

- Commits authored by `dependabot[bot]`, `dependabot-preview[bot]` or `github-actions[bot]`
- Bulk update PRs made by `ErikSchierboom` or `kytrinx` to update the track

## Tracking

https://github.com/exercism/v3-launch/issues/24

---

cc @adriankumpf, @alarregoity, @amencarini, @ananthamapod, @andrewsardone, @andrewtimberlake, @angelikatyborska, @AnilRedshift, @austinlyons, @awochna, @bartj3, @bernardoamc, @Billiam, @Br1ght0ne, @brianriley, @brianvanburken, @britto, @Bscruz19, @bunnymatic, @cbliard, @cetinajero, @cezarykosko, @chriseyre2000, @ChristianTovar, @chriszimmerman, @CoderDennis, @Cohen-Carlisle, @crazymykl, @dalexj, @DanCouper, @dantswain, @davbre, @daveyarwood, @davinlagerroos, @devonestes, @digitalronin, @DoggettCK, @doncruse, @drueck, @ebiven, @elasticdog, @etrepum, @fhdhsni, @flyrry, @frerich, @fxn, @George-Hudson, @ggpasqualino, @ghajba, @glennular, @gmile, @gvaughn, @hanmd82, @henrik, @herminiotorres, @jason-kerney, @jedschneider, @jeremy-w, @jimmbraddock, @jinyeow, @johnsyweb, @jwworth, @korbin, @koriroys, @krishandley, @kronn, @kytrinyx, @lancehalvorsen, @leikind, @lex57ukr, @lpil, @lsimoneau, @lucasprag, @MarcosX, @markijbema, @martinsvalin, @meadsteve, @MeerKatDev, @mexicat, @mhinz, @mlevesquedion, @mlopes, @montague, @moxley, @nathanchere, @nathanielknight, @neenjaw, @NickNeck, @nilvon9wo, @nimser, @NobbZ, @nwshane, @olleolleolle, @parkerl, @patrickgombert, @PatrickMcSweeny, @percygrunwald, @petehuang, @petertseng, @pminten, @rawkode, @rsslldnphy, @rubysolo, @rud, @ryanzidago, @Scientifica96, @screamingjungle, @seeflanigan, @sngeth, @sotojuan, @stfnsr, @Teapane, @tejasbubane, @Thrillberg, @tjcelaya, @Tonkpils, @toriejw, @tsuka, @Tuxified, @v-kolesnikov, @veelenga, @victorlcampos, @victorpre, @vladimir-tikhonov, @waiting-for-dev, @weapp, @workingjubilee, @ybod, @zporter as you are referenced in this PR

### CODE REVIEW

**Comment by angelikatyborska** [comment]
(cc @gmile, @gvaughn, @hanmd82, @henrik, @herminiotorres, @jason-kerney, @jedschneider, @jeremy-w, @jimmbraddock, @jinyeow, @johnsyweb, @jwworth, @korbin, @koriroys, @krishandley, @kronn, @kytrinyx, @lancehalvorsen, @leikind, @lex57ukr, @lpil, @lsimoneau, @lucasprag, @MarcosX, @markijbema, @martinsvalin, @meadsteve, @MeerKatDev, @mexicat, @mhinz, @mlevesquedion, @mlopes, @montague, @moxley, @nathanchere, @nathanielknight, @neenjaw, @NickNeck, @nilvon9wo, @nimser, @NobbZ, @nwshane, @olleolleolle, @parkerl, @patrickgombert, @PatrickMcSweeny, @percygrunwald, @petehuang, @petertseng, @pminten as you are referenced in this PR)

**Comment by angelikatyborska** [comment]
(cc @rawkode, @rsslldnphy, @rubysolo, @rud, @ryanzidago, @Scientifica96, @screamingjungle, @seeflanigan, @sngeth, @sotojuan, @stfnsr, @Teapane, @tejasbubane, @Thrillberg, @tjcelaya, @Tonkpils, @toriejw, @tsuka, @Tuxified, @v-kolesnikov, @veelenga, @victorlcampos, @victorpre, @vladimir-tikhonov, @waiting-for-dev, @weapp, @workingjubilee, @ybod, @zporter as you are referenced in this PR)

**Comment by angelikatyborska** [comment]
(tagging people split into multiple comments because there seems to be a limit to how many people can be tagged in a single comment)

**Review by lpil — APPROVED** [review]

**Review by angelikatyborska — APPROVED** [review]

**Review by johnsyweb — APPROVED** [review]
❤️

File: exercises/practice/binary/.meta/config.json [hunk 1]
```diff
@@ -1,6 +1,24 @@
...
+    "andrewsardone",
+    "angelikatyborska",
+    "Cohen-Carlisle",
+    "dalexj",
+    "devonestes",
+    "jinyeow",
+    "kytrinyx",
+    "lpil",
+    "neenjaw",
+    "NobbZ",
```
Comment (NobbZ) [inline exercises/practice/binary/.meta/config.json:16]: I can't find myself in the commit history, can someone point me to the location? Even though I can't remember contributing, the combination from language and exercise makes it likely I indeed did :D

Comment (angelikatyborska) [reply exercises/practice/binary/.meta/config.json:16]: I found at least three commits of yours for that exercise:
https://github.com/exercism/elixir/commit/b0d56986e140db7ac55d59082cfba7235fac43d3
https://github.com/exercism/elixir/commit/70d4e7c6a1926ca339ff5092c2abef69c000519f
https://github.com/exercism/elixir/commit/0eb33b7da1f7b35dc27703327cc10546e77f48d6

**Review by Tuxified — APPROVED** [review]

**Review by jeremy-w — APPROVED** [review]

**Review by Teapane — COMMENTED** [review]
LGTM. Thanks for doing this!

**Review by gvaughn — APPROVED** [review]

**Review by kronn — APPROVED** [review]

File: exercises/practice/gigasecond/.meta/config.json [hunk 2]
```diff
@@ -1,6 +1,24 @@
 {
   "blurb": "Given a moment, determine the moment that would be after a gigasecond has passed.",
-  "authors": [],
+  "authors": [
+    "rubysolo"
+  ],
+  "contributors": [
+    "andrewsardone",
+    "angelikatyborska",
+    "Billiam",
```
Comment (Billiam) [inline exercises/practice/gigasecond/.meta/config.json:9]: My contribution was a whitespace-only change, and should probably not be included here. Thanks though!

**Review by frerich — APPROVED** [review]

**Review by tejasbubane — APPROVED** [review]

**Review by chriszimmerman — APPROVED** [review]

**Review by fxn — APPROVED** [review]

**Review by gmile — APPROVED** [review]

**Review by amencarini — APPROVED** [review]

**Review by ChristianTovar — APPROVED** [review]

**Review by herminiotorres — APPROVED** [review]

**Review by bunnymatic — APPROVED** [review]

**Review by brianvanburken — APPROVED** [review]

**Review by rubysolo — APPROVED** [review]

**Review by britto — APPROVED** [review]

**Comment by angelikatyborska** [comment]
7 days have passed and a lot of people expressed their approval by either leaving a review or an emoji reaction :) I think it's time to merge!

In case somebody didn't have a chance to review it yet and finds themselves wishing to make changes about where they themselves are mentioned, feel free to open a new PR with your desired changes.

**Pull request merged**